scipy/spatial/_lloyd_algorithm.py

```diff
@@ -0,0 +1,241 @@
+"""Lloyd-Max algorithm."""
+from __future__ import annotations
+
+import numpy as np
+from typing import (
+    Optional,
+    TYPE_CHECKING,
+)
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from scipy._lib._util import (
+        DecimalNumber, IntNumber
+    )
+
+from scipy.spatial import distance, Voronoi
+
+
+__all__ = ['lloyd_centroidal_voronoi_tessellation']
+
+
+def _l1_norm(points: np.ndarray) -> float:
+    return distance.pdist(points, 'cityblock').min()
+
+
+def _lloyd_centroidal_voronoi_tessellation(
+        points: np.ndarray,
+        decay: float,
+        qhull_options: str,
+) -> np.ndarray:
+    """Lloyd-Max algorithm iteration.
+
+    Based on the implementation of Stéfan van der Walt:
+
+    https://github.com/stefanv/lloyd
+
+    which is:
+
+        Copyright (c) 2021-04-21 Stéfan van der Walt
+        https://github.com/stefanv/lloyd
+        MIT License
+
+    Parameters
+    ----------
+    points : array_like (n, d)
+        The points to iterate on.
+    decay : float
+        Relaxation decay. A positive value would move the points toward
+        their centroid, and negative value would move them away.
+        1 would move the points to their centroid.
+    qhull_options : str
+        Additional options to pass to Qhull. See Qhull manual
+        for details. (Default: "Qbb Qc Qz Qj Qx" for ndim > 4 and
+        "Qbb Qc Qz Qj" otherwise.)
+
+    Returns
+    -------
+    points : array_like (n, d)
+        The points after an iteration of Lloyd's algorithm.
+
+    """
+    new_points = np.empty_like(points)
+
+    voronoi = Voronoi(points, qhull_options=qhull_options)
+
+    for ii, idx in enumerate(voronoi.point_region):
+        # the region is a series of indices into self.voronoi.vertices
+        # remove point at infinity, designated by index -1
+        region = [i for i in voronoi.regions[idx] if i != -1]
+
+        # get the vertices for this region
+        verts = voronoi.vertices[region]
+
+        # clipping would be wrong, we need to intersect
+        # verts = np.clip(verts, 0, 1)
+
+        # move points towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
+        centroid = np.mean(verts, axis=0)
+        new_points[ii] = points[ii] + (centroid - points[ii]) * decay
+
+    # only update points to centroid within the region
+    is_valid = np.all(np.logical_and(new_points >= 0, new_points <= 1), axis=1)
+    points[is_valid] = new_points[is_valid]
+
+    return points
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        points: npt.ArrayLike,
+        *,
+        tol: DecimalNumber = 1e-5,
+        maxiter: IntNumber = 10,
+        decay: Optional[npt.ArrayLike] = None,
+        qhull_options: Optional[str] = None,
+) -> np.ndarray:
+    """Approximate Centroidal Voronoi Tessellation.
+
+    Perturb points in :math:`[0, 1]^d` using Lloyd-Max algorithm.
```

Comment:
Let's move the :math:`[0, 1]^d` requirement down to the `points` parameter. If you want to specify that it moves points in  Nd, that sounds good, but the unit hypercube thing is a detail.

Reply:
Done.